[feature] Add a QgsAppGpsTools class to expose QGIS GPS Tools to Python plugins via QgisInterface

gacarrillor opened the pull request:
This PR exposes QGIS GPS tools via a brand new `QgsAppGpsTools` class, which can be accessed via `QgisInterface::gpsTools()`, so that plugins can have more control on the QGIS GPS tracking session.

To avoid the need of adding new custom methods to `QgisApp` using the `QgsAppGpsDigitizing` object, we expose the object itself via `QgisApp::gpsDigitizing()`.

From QGIS 3.44 on, plugins can create a vector layer feature from the current GPS track (something only available via GUI up to now):

    iface.gpsTools().createFeatureFromGpsTrack()

as well as set the GPS track line symbol, changing the corresponding setting in `QgsOptions` and updating the eventual current GPS track geometry accordingly:

    iface.gpsTools().setGpsTrackLineSymbol( line_symbol )


Thanks to @nyalldawson for the suggestions about optimizing these changes!

--------------

Funded by the Danish Nature Agency, [Naturstyrelsen](https://naturstyrelsen.dk/).

gacarrillor committed:
[feature] Expose QgsAppGpsDigitizing::createFeature() through iface, so that Python plugins can have more control on the GPS tracking session

gacarrillor committed:
[feature] Allow Python plugins to set the GPS track line symbol (changing its corresponding setting in GPS options) and update the symbol of an eventual current GPS track

nyalldawson commented:
I'd like to review this, but will need some time

3nids commented:
@nyalldawson do you still want to review? Otherwise, I'll merge it at the end of the week.

nyalldawson commented:
Ok -- my concern here is cluttering up the already massive QgisInterface class. I'm on a crusade to avoid the ongoing practice of just throwing stuff into QgisInterface/QgisApp because it's convenient, and instead change to using nicely designed classes  for new API.

So, I would prefer that these methods are instead placed into a new QgsGpsToolsInterface abstract class, and then a single:

    virtual QgsGpsToolsInterface* gpsTools() = 0;

method in QgisInterface. Then there'd be a QgsAppGpsTools class made which implements this new interface.

gacarrillor committed:
Refactor of GPS tools exposed via iface:

Introduce abstract QgsGpsToolsInterface and its subclass QgsAppGpsTools, which can be accessed via iface.gpsTools()
Deprecate existent GPS method in iface and move it to GPS tools.

nyalldawson commented on src/core/gps/qgsgpstoolsinterface.h:
```diff
@@ -0,0 +1,62 @@
+ * \brief Abstract interface class for the QGIS GPS tools.
+ * \since QGIS 3.44
+ */
+class CORE_EXPORT QgsGpsToolsInterface : public QObject
```
Can you move this to gui, like QgisInterface is?

nyalldawson replied:
At this stage this doesn't need to be a QObject, let's avoid:

```suggestion
class CORE_EXPORT QgsGpsToolsInterface
```

nyalldawson commented on src/app/gps/qgsappgpstools.cpp:
```diff
@@ -0,0 +1,35 @@
+
+void QgsAppGpsTools::setGpsTrackLineSymbol( QgsLineSymbol *symbol )
+{
+  QgisApp::instance()->setGpsTrackLineSymbol( symbol );
```
Could we call QgsAppGpsDigitizing directly here, avoiding the need for new methods in QgisApp?

```suggestion
  QgsAppGpsDigitizing::setGpsTrackLineSymbol( symbol );
```

gacarrillor replied:
That would be good indeed, but we'd still need to call:

`mGpsDigitizing->updateTrackAppearance();`

Which is non-static. Same for

`mGpsDigitizing->createFeature();`

in the method above.

What would be your suggestion, could we expose `mGpsDigitizing` from `QgisApp` instead of the new methods?

nyalldawson replied:
Yes, that sounds like a good idea!

nyalldawson commented on src/app/qgisappinterface.h:
```diff
@@ -333,6 +336,9 @@ class APP_EXPORT QgisAppInterface : public QgisInterface
     QgsAppPluginManagerInterface pluginManagerIface;
+
+    //! Pointer to the GpsToolsInterface object
+    QgsAppGpsTools gpsToolsIface;
```
```suggestion
    QgsAppGpsTools mGpsToolsIface;
```

nyalldawson commented on src/core/gps/qgsgpstoolsinterface.h:
```diff
@@ -0,0 +1,62 @@
+{
+    Q_OBJECT
+  public:
+    QgsGpsToolsInterface() = default;
```
```suggestion
    virtual ~QgsGpsToolsInterface() = default;
```

nyalldawson commented on src/gui/qgisinterface.h:
```diff
@@ -1503,9 +1509,10 @@ class GUI_EXPORT QgisInterface : public QObject
      * Any existing GPS connection used by the widget will be disconnect and replaced with this connection. The connection
      * is automatically registered within the QgsApplication::gpsConnectionRegistry().
      *
+     * \deprecated QGIS 3.44. Use the method from QgisAppInterface::gpsTools() instead.
```
```suggestion
     * \deprecated QGIS 3.44. Use the method from gpsTools() instead.
```

nyalldawson commented:
Thanks @gacarrillor  -- this is looking good!

nyalldawson commented:
Thanks @gacarrillor -- this is looking good!

gacarrillor commented:
(I'll finish this during the next couple of days)

gacarrillor commented:
(I had prepared this note yesterday but you were too fast on the review :) )

Thanks @nyalldawson for the initial suggestion and the follow-up review here. We've addressed them all.

Note we also deprecated and moved the existent `QgisInterface::setGpsPanelConnection()` method to the new `QgsAppGpsTools()` class.

nyalldawson commented on src/gui/qgsgpstoolsinterface.cpp:
```diff
@@ -0,0 +1,16 @@
+ *                                                                         *
+ ***************************************************************************/
+#include "qgsgpstoolsinterface.h"
+#include "moc_qgsgpstoolsinterface.cpp"
```
```suggestion
```

nyalldawson committed:
Update src/gui/qgsgpstoolsinterface.cpp
